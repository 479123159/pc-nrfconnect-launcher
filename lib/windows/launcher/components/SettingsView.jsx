/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

import React from 'react';
import PropTypes from 'prop-types';
import { Checkbox } from 'react-bootstrap';
import moment from 'moment';

class SettingsView extends React.Component {
    constructor() {
        super();
        this.onCheckUpdatesAtStartupChanged = this.onCheckUpdatesAtStartupChanged.bind(this);
        this.onTriggerUpdateCheckClicked = this.onTriggerUpdateCheckClicked.bind(this);
    }

    componentDidMount() {
        this.props.onMount();
    }

    onCheckUpdatesAtStartupChanged(event) {
        const { onCheckUpdatesAtStartupChanged } = this.props;
        onCheckUpdatesAtStartupChanged(event.target.checked);
    }

    onTriggerUpdateCheckClicked() {
        this.props.onTriggerUpdateCheck();
    }

    render() {
        const {
            isLoading,
            shouldCheckForUpdatesAtStartup,
            isCheckingForUpdates,
            lastUpdateCheckDate,
        } = this.props;

        const checkButtonText = isCheckingForUpdates ? 'Checking...' : 'Check for updates now';

        return !isLoading ? (
            <div>
                <div className="core-settings-section">
                    <h4>Updates</h4>
                    {
                        lastUpdateCheckDate ?
                            <p>Last update check performed: { moment(lastUpdateCheckDate).format('YYYY-MM-DD HH:mm:ss') }</p> :
                            null
                    }
                    <div className="core-settings-update-check-controls">
                        <Checkbox
                            checked={shouldCheckForUpdatesAtStartup}
                            onChange={this.onCheckUpdatesAtStartupChanged}
                        >
                            Check for updates at startup
                        </Checkbox>
                        <button
                            title={checkButtonText}
                            className="btn btn-primary core-btn"
                            onClick={this.onTriggerUpdateCheckClicked}
                            disabled={isCheckingForUpdates}
                        >
                            <span className="glyphicon glyphicon-refresh" />
                            <span className="core-btn-text">
                                { checkButtonText }
                            </span>
                        </button>
                    </div>
                </div>
            </div>
        ) : <div />;
    }
}

SettingsView.propTypes = {
    onMount: PropTypes.func,
    isLoading: PropTypes.bool.isRequired,
    shouldCheckForUpdatesAtStartup: PropTypes.bool.isRequired,
    isCheckingForUpdates: PropTypes.bool.isRequired,
    onCheckUpdatesAtStartupChanged: PropTypes.func.isRequired,
    onTriggerUpdateCheck: PropTypes.func.isRequired,
    lastUpdateCheckDate: PropTypes.instanceOf(Date),
};

SettingsView.defaultProps = {
    onMount: () => {},
    lastUpdateCheckDate: null,
};

export default SettingsView;
