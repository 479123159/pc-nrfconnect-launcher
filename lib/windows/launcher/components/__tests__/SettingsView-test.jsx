/* Copyright (c) 2015 - 2017, Nordic Semiconductor ASA
 *
 * All rights reserved.
 *
 * Use in source and binary forms, redistribution in binary form only, with
 * or without modification, are permitted provided that the following conditions
 * are met:
 *
 * 1. Redistributions in binary form, except as embedded into a Nordic
 *    Semiconductor ASA integrated circuit in a product or a software update for
 *    such product, must reproduce the above copyright notice, this list of
 *    conditions and the following disclaimer in the documentation and/or other
 *    materials provided with the distribution.
 *
 * 2. Neither the name of Nordic Semiconductor ASA nor the names of its
 *    contributors may be used to endorse or promote products derived from this
 *    software without specific prior written permission.
 *
 * 3. This software, with or without modification, must only be used with a Nordic
 *    Semiconductor ASA integrated circuit.
 *
 * 4. Any software provided in binary form under this license must not be reverse
 *    engineered, decompiled, modified and/or disassembled.
 *
 * THIS SOFTWARE IS PROVIDED BY NORDIC SEMICONDUCTOR ASA "AS IS" AND ANY EXPRESS OR
 * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF
 * MERCHANTABILITY, NONINFRINGEMENT, AND FITNESS FOR A PARTICULAR PURPOSE ARE
 * DISCLAIMED. IN NO EVENT SHALL NORDIC SEMICONDUCTOR ASA OR CONTRIBUTORS BE LIABLE
 * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
 * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
 * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
 * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR
 * TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
 * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 */

/* eslint-disable import/first */

// Do not render react-bootstrap components in tests
jest.mock('react-bootstrap', () => ({
    Modal: 'Modal',
    Button: 'Button',
    ModalHeader: 'ModalHeader',
    ModalFooter: 'ModalFooter',
    ModalBody: 'ModalBody',
    ModalTitle: 'ModalTitle',
    Checkbox: 'Checkbox',
}));

import React from 'react';
import renderer from 'react-test-renderer';
import SettingsView from '../SettingsView';

describe('SettingsView', () => {
    it('should render empty div while loading', () => {
        expect(renderer.create(
            <SettingsView
                isLoading
                shouldCheckForUpdatesAtStartup
                isCheckingForUpdates={false}
                onTriggerUpdateCheck={() => {}}
                onCheckUpdatesAtStartupChanged={() => {}}
                onHideUpdateCheckCompleteDialog={() => {}}
            />,
        )).toMatchSnapshot();
    });

    it('should render with check for updates enabled', () => {
        expect(renderer.create(
            <SettingsView
                isLoading={false}
                shouldCheckForUpdatesAtStartup
                isCheckingForUpdates={false}
                onTriggerUpdateCheck={() => {}}
                onCheckUpdatesAtStartupChanged={() => {}}
                onHideUpdateCheckCompleteDialog={() => {}}
            />,
        )).toMatchSnapshot();
    });

    it('should render with check for updates disabled', () => {
        expect(renderer.create(
            <SettingsView
                isLoading={false}
                shouldCheckForUpdatesAtStartup={false}
                isCheckingForUpdates={false}
                onTriggerUpdateCheck={() => {}}
                onCheckUpdatesAtStartupChanged={() => {}}
                onHideUpdateCheckCompleteDialog={() => {}}
            />,
        )).toMatchSnapshot();
    });

    it('should render when checking for updates', () => {
        expect(renderer.create(
            <SettingsView
                isLoading={false}
                shouldCheckForUpdatesAtStartup={false}
                isCheckingForUpdates
                onTriggerUpdateCheck={() => {}}
                onCheckUpdatesAtStartupChanged={() => {}}
                onHideUpdateCheckCompleteDialog={() => {}}
            />,
        )).toMatchSnapshot();
    });

    it('should render with last update check date', () => {
        expect(renderer.create(
            <SettingsView
                isLoading={false}
                shouldCheckForUpdatesAtStartup={false}
                isCheckingForUpdates={false}
                lastUpdateCheckDate={new Date(2017, 1, 3, 13, 41, 36, 20)}
                onTriggerUpdateCheck={() => {}}
                onCheckUpdatesAtStartupChanged={() => {}}
                onHideUpdateCheckCompleteDialog={() => {}}
            />,
        )).toMatchSnapshot();
    });

    it('should render check for updates completed, with updates available', () => {
        expect(renderer.create(
            <SettingsView
                isLoading={false}
                shouldCheckForUpdatesAtStartup={false}
                isCheckingForUpdates={false}
                lastUpdateCheckDate={new Date(2017, 1, 3, 13, 41, 36, 20)}
                onTriggerUpdateCheck={() => {}}
                onCheckUpdatesAtStartupChanged={() => {}}
                onHideUpdateCheckCompleteDialog={() => {}}
                isUpdateCheckCompleteDialogVisible
                isAppUpdateAvailable
            />,
        )).toMatchSnapshot();
    });

    it('should render check for updates completed, with everything up to date', () => {
        expect(renderer.create(
            <SettingsView
                isLoading={false}
                shouldCheckForUpdatesAtStartup={false}
                isCheckingForUpdates={false}
                lastUpdateCheckDate={new Date(2017, 1, 3, 13, 41, 36, 20)}
                onTriggerUpdateCheck={() => {}}
                onCheckUpdatesAtStartupChanged={() => {}}
                onHideUpdateCheckCompleteDialog={() => {}}
                isUpdateCheckCompleteDialogVisible
                isAppUpdateAvailable={false}
            />,
        )).toMatchSnapshot();
    });
});
