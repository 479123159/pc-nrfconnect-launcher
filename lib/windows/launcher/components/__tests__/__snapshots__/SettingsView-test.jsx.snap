// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SettingsView should render check for updates completed, with everything up to date 1`] = `
<div>
  <div
    className="core-settings-section"
  >
    <h4>
      Updates
    </h4>
    <p>
      Last update check performed: 
      2017-02-03 13:41:36
    </p>
    <div
      className="core-settings-update-check-controls"
    >
      <Checkbox
        checked={false}
        onChange={[Function]}
      >
        Check for updates at startup
      </Checkbox>
      <Button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Check for updates now"
      >
        <span
          className="glyphicon glyphicon-refresh"
        />
        <span
          className="core-btn-text"
        >
          Check for updates now
        </span>
      </Button>
    </div>
  </div>
  <Modal
    backdrop={false}
    onHide={null}
    show={true}
  >
    <ModalHeader
      closeButton={true}
    >
      <ModalTitle>
        Update check completed
      </ModalTitle>
    </ModalHeader>
    <ModalBody>
      <p>
        All apps are up to date.
      </p>
    </ModalBody>
    <ModalFooter>
       
      <Button
        bsStyle="primary"
        className="core-btn"
        disabled={false}
        onClick={[Function]}
      >
        OK
      </Button>
    </ModalFooter>
  </Modal>
  <div
    className="core-settings-section"
    onDragEnter={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <h4>
      Extra app sources
    </h4>
    <table
      className="core-settings-sources"
    >
      <tbody />
    </table>
    <div
      className="core-settings-sources-controls"
    >
      <Button
        className="btn btn-primary core-btn"
        onClick={[Function]}
      >
        <span
          className="glyphicon glyphicon-plus"
        />
        <span
          className="core-btn-text"
        >
          Add source
        </span>
      </Button>
    </div>
  </div>
  <ConfirmRemoveSourceDialog />
</div>
`;

exports[`SettingsView should render check for updates completed, with updates available 1`] = `
<div>
  <div
    className="core-settings-section"
  >
    <h4>
      Updates
    </h4>
    <p>
      Last update check performed: 
      2017-02-03 13:41:36
    </p>
    <div
      className="core-settings-update-check-controls"
    >
      <Checkbox
        checked={false}
        onChange={[Function]}
      >
        Check for updates at startup
      </Checkbox>
      <Button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Check for updates now"
      >
        <span
          className="glyphicon glyphicon-refresh"
        />
        <span
          className="core-btn-text"
        >
          Check for updates now
        </span>
      </Button>
    </div>
  </div>
  <Modal
    backdrop={false}
    onHide={null}
    show={true}
  >
    <ModalHeader
      closeButton={true}
    >
      <ModalTitle>
        Update check completed
      </ModalTitle>
    </ModalHeader>
    <ModalBody>
      <p>
        One or more updates are available. Go to the Add/remove apps screen to upgrade.
      </p>
    </ModalBody>
    <ModalFooter>
       
      <Button
        bsStyle="primary"
        className="core-btn"
        disabled={false}
        onClick={[Function]}
      >
        OK
      </Button>
    </ModalFooter>
  </Modal>
  <div
    className="core-settings-section"
    onDragEnter={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <h4>
      Extra app sources
    </h4>
    <table
      className="core-settings-sources"
    >
      <tbody />
    </table>
    <div
      className="core-settings-sources-controls"
    >
      <Button
        className="btn btn-primary core-btn"
        onClick={[Function]}
      >
        <span
          className="glyphicon glyphicon-plus"
        />
        <span
          className="core-btn-text"
        >
          Add source
        </span>
      </Button>
    </div>
  </div>
  <ConfirmRemoveSourceDialog />
</div>
`;

exports[`SettingsView should render empty div while loading 1`] = `<div />`;

exports[`SettingsView should render when checking for updates 1`] = `
<div>
  <div
    className="core-settings-section"
  >
    <h4>
      Updates
    </h4>
    <div
      className="core-settings-update-check-controls"
    >
      <Checkbox
        checked={false}
        onChange={[Function]}
      >
        Check for updates at startup
      </Checkbox>
      <Button
        className="btn btn-primary core-btn"
        disabled={true}
        onClick={[Function]}
        title="Checking..."
      >
        <span
          className="glyphicon glyphicon-refresh"
        />
        <span
          className="core-btn-text"
        >
          Checking...
        </span>
      </Button>
    </div>
  </div>
  <div
    className="core-settings-section"
    onDragEnter={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <h4>
      Extra app sources
    </h4>
    <table
      className="core-settings-sources"
    >
      <tbody />
    </table>
    <div
      className="core-settings-sources-controls"
    >
      <Button
        className="btn btn-primary core-btn"
        onClick={[Function]}
      >
        <span
          className="glyphicon glyphicon-plus"
        />
        <span
          className="core-btn-text"
        >
          Add source
        </span>
      </Button>
    </div>
  </div>
  <ConfirmRemoveSourceDialog />
</div>
`;

exports[`SettingsView should render with check for updates disabled 1`] = `
<div>
  <div
    className="core-settings-section"
  >
    <h4>
      Updates
    </h4>
    <div
      className="core-settings-update-check-controls"
    >
      <Checkbox
        checked={false}
        onChange={[Function]}
      >
        Check for updates at startup
      </Checkbox>
      <Button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Check for updates now"
      >
        <span
          className="glyphicon glyphicon-refresh"
        />
        <span
          className="core-btn-text"
        >
          Check for updates now
        </span>
      </Button>
    </div>
  </div>
  <div
    className="core-settings-section"
    onDragEnter={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <h4>
      Extra app sources
    </h4>
    <table
      className="core-settings-sources"
    >
      <tbody />
    </table>
    <div
      className="core-settings-sources-controls"
    >
      <Button
        className="btn btn-primary core-btn"
        onClick={[Function]}
      >
        <span
          className="glyphicon glyphicon-plus"
        />
        <span
          className="core-btn-text"
        >
          Add source
        </span>
      </Button>
    </div>
  </div>
  <ConfirmRemoveSourceDialog />
</div>
`;

exports[`SettingsView should render with check for updates enabled 1`] = `
<div>
  <div
    className="core-settings-section"
  >
    <h4>
      Updates
    </h4>
    <div
      className="core-settings-update-check-controls"
    >
      <Checkbox
        checked={true}
        onChange={[Function]}
      >
        Check for updates at startup
      </Checkbox>
      <Button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Check for updates now"
      >
        <span
          className="glyphicon glyphicon-refresh"
        />
        <span
          className="core-btn-text"
        >
          Check for updates now
        </span>
      </Button>
    </div>
  </div>
  <div
    className="core-settings-section"
    onDragEnter={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <h4>
      Extra app sources
    </h4>
    <table
      className="core-settings-sources"
    >
      <tbody />
    </table>
    <div
      className="core-settings-sources-controls"
    >
      <Button
        className="btn btn-primary core-btn"
        onClick={[Function]}
      >
        <span
          className="glyphicon glyphicon-plus"
        />
        <span
          className="core-btn-text"
        >
          Add source
        </span>
      </Button>
    </div>
  </div>
  <ConfirmRemoveSourceDialog />
</div>
`;

exports[`SettingsView should render with last update check date 1`] = `
<div>
  <div
    className="core-settings-section"
  >
    <h4>
      Updates
    </h4>
    <p>
      Last update check performed: 
      2017-02-03 13:41:36
    </p>
    <div
      className="core-settings-update-check-controls"
    >
      <Checkbox
        checked={false}
        onChange={[Function]}
      >
        Check for updates at startup
      </Checkbox>
      <Button
        className="btn btn-primary core-btn"
        disabled={false}
        onClick={[Function]}
        title="Check for updates now"
      >
        <span
          className="glyphicon glyphicon-refresh"
        />
        <span
          className="core-btn-text"
        >
          Check for updates now
        </span>
      </Button>
    </div>
  </div>
  <div
    className="core-settings-section"
    onDragEnter={[Function]}
    onDragOver={[Function]}
    onDrop={[Function]}
  >
    <h4>
      Extra app sources
    </h4>
    <table
      className="core-settings-sources"
    >
      <tbody />
    </table>
    <div
      className="core-settings-sources-controls"
    >
      <Button
        className="btn btn-primary core-btn"
        onClick={[Function]}
      >
        <span
          className="glyphicon glyphicon-plus"
        />
        <span
          className="core-btn-text"
        >
          Add source
        </span>
      </Button>
    </div>
  </div>
  <ConfirmRemoveSourceDialog />
</div>
`;
